smolyay/basis.py

```diff
@@ -328,141 +262,137 @@ def derivative(self, x):
         Returns
         -------
         float
-            Value of the derivative of Chebyshev polynomials of first kind.
+            Value of the derivative of Chebyshev polynomials of second kind.
 
         Raises
         ------
         ValueError
-            if input is outside the domain [-1,1].
+            if input is outside the domain [-1, 1].
         """
-        if numpy.any(numpy.greater(x, 1)) or numpy.any(numpy.less(x, -1)):
-            raise ValueError("Input is outside the domain [-1,1]")
-        if self._derivative_polynomial is None:
-            self._derivative_polynomial = ChebyshevFirstKind(self._n+1)
+        if self.outside_domain(x):
+            raise ValueError("Input is outside the domain [-1, 1]")
         x = numpy.asarray(x)
         y = numpy.zeros(x.shape)
         flag2 = x == 1
         flag3 = x == -1
         flag1 = ~(flag2 | flag3)
-        y[flag1] =  ((self._n+1)*self._derivative_polynomial(x[flag1]) -
-                     x[flag1]*self(x[flag1]))/(x[flag1]**2-1)
-        y[flag2] = self._n*(self._n + 1)*(self._n + 2)/3
-        y[flag3] = ((-1)**(self._n+1))*self._n*(self._n + 1)*(self._n + 2)/3
+        y[flag1] = (
+            (self._n + 1) * scipy.special.eval_chebyt(self._n + 1, x[flag1])
+            - x[flag1] * self(x[flag1])
+        ) / (x[flag1] ** 2 - 1)
+        y[flag2] = self._n * (self._n + 1) * (self._n + 2) / 3
+        y[flag3] = ((-1) ** (self._n + 1)) * self._n * (self._n + 1) * (self._n + 2) / 3
```

Comment:
We should recycle the evaluation on line 283 in line 284, since 284 is 283 times a minus sign. Instead of raising -1 to the power of n, you should instead check whether n is odd (-1) or even (+1) to determine the sign.

Reply:
By replacing -1 with numpy.sign(x[flag2]) and flag2 = numpy.abs(x) == 1, I can get the entire evaluation in one equation. Using modulo operations to check for even or odd results in either a 0 or a 1, which requires more work to make a -1 and 1, and trying to use modulo ended up being more cumbersome than raising 1 or -1 to some power.